src/core/qgsmapsettings.cpp

```diff
@@ -496,11 +496,11 @@ QgsPointXY QgsMapSettings::layerToMapCoordinates( const QgsMapLayer *layer, QgsP
   return point;
 }
 
-QgsPoint QgsMapSettings::layerToMapCoordinates( const QgsMapLayer *layer, QgsPoint point ) const
+QgsPoint QgsMapSettings::layerToMapCoordinates( const QgsMapLayer *layer, const QgsPoint &point ) const
 {
   double x = point.x();
   double y = point.y();
-  double z = point.z();
+  double z = !std::isnan( point.z() ) ? point.z() : 0.0;
```

Comment:
I think we might need to handle this at a lower level -- in QgsCoordinateTransform::transformCoords itself. Otherwise we're likely to have other circumstances where a QgsPoint is transformed which will exhibit similar issues.

What we'd need to do is change these tests:

```
    // proj_errno is sometimes not an accurate method to test for transform failures - so we need to
    // manually scan for nan values
    if ( std::any_of( x, x + numPoints, []( double v ) { return std::isinf( v ); } )
    || std::any_of( y, y + numPoints, []( double v ) { return std::isinf( v ); } )
    || std::any_of( z, z + numPoints, []( double v ) { return std::isinf( v ); } ) )
    {
      actualRes = 1;
    }
```

To also test whether any output nan values are NOT nan values in the xprev/yprev/zprev arrays. That should fix the ultimate underlying issue causing this behavior.

Reply:
Right, I suspected as much.